### PR TITLE
build GitHub plugin

### LINKED ISSUE #907: Create Github plugin
sub of #898 

**what**

A plugin that requests data about Github repositories using the Github API using personal authorization tokens stored in environment variables.

**context**

Code storage and distribution is an important component of quantifying the environmental impacts of software projects. Github is the industry standard platform for hosting code. They expose an API that serves repository metrics incuding repo size and frequency of clones etc that can be used to assess the environmental impacts of the project storage and distribution.

Specifically, the plugin should return the following:

| value                              | API endpoint                                                                                                                                                                                    | Docs link                                                             |
| ---------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------- |
| number of time the repo was clones | `curl -L \ -H "Accept: application/vnd.github+json" \ -H "Authorization: Bearer <YOUR-TOKEN>" \ -H "X-GitHub-Api-Version: 2022-11-28" \ https://api.github.com/repos/OWNER/REPO/traffic/clones` | https://docs.github.com/en/rest/metrics/traffic?apiVersion=2022-11-28 |
| repo size in kb                    | `curl -L \ -H "Accept: application/vnd.github+json" \ -H "Authorization: Bearer <YOUR-TOKEN>" \ -H "X-GitHub-Api-Version: 2022-11-28" \ https://api.github.com/orgs/ORG/repos`                  | https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28     |

**SoW**

- create plugin that returns the requested data from theGithub API
- create in dedicated IF repository (not a builtin)
- add plugin to Explorer

**Acceptance criteria**

- [x] The plugin retrieves `size` and `clones` data from the Github API. `size` should be converted from `kb` to `GB` for consistency with other IF plugins.

**Note** that the traffic metrics only extend back as far as the most recent four weeks. You can request the total for any of those weeks or the whole preceding month. I think the right way to handle this in IF is to offer two options: last week or last month. 

**GIVEN** the plugin works as expected and I have a valid auth token in `.env`
**WHEN** I run the following manifest:

```yaml
name: github demo
description:
tags:
initialize:
  plugins:
    github-plugin:
      method: GithubAPI
      path: github-api
      config:
        repo-name: Green-Software-Foundation/if
        time: 'week' # can be week or month
tree:
  children:
    github:
      pipeline:
        observe:
         - github-plugin
      config:
      defaults:
      inputs:
        - timestamp: 2023-08-06T00:00
          duration: 2419200
```

**Then** I get the following output (the last week's clones and the current repo size converted from kb to GB:

```yaml
name: github demo
description:
tags:
initialize:
  plugins:
    github-plugin:
      method: GithubAPI
      path: github-api
      config:
        repo-name: Green-Software-Foundation/if
        time: 'week' # can be week or month
tree:
  children:
    github:
      pipeline:
        observe:
         - github-plugin
      config:
      defaults:
      inputs:
        - timestamp: 2023-08-06T00:00
          duration: 2419200
      outputs:
        - timestamp: 2023-08-06T00:00
          duration: 2419200
          size: 0.0017
          clones: 34
      
```

- [x] The plugin is available on https://explorer.if.greensoftware.foundation

Note to be listed it has to meet some basic standards for documentation and repository structure.

### PR DESCRIPTION
**Pull request by manushak**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+```sh
+if-run -m <path-to-your-manifest>
+```
+
+## References
+
+The plugin simply grabs data for a given repository from the github.com API.
+
+- To calculate `clones`, the plugin uses [this](https://docs.github.com/en/rest/metrics/traffic) endpoint.
```
Comment (narekhovhannisyan) [inline README.md]:
```suggestion
- To calculate `clones`, the plugin uses [traffic](https://docs.github.com/en/rest/metrics/traffic) endpoint.
```

File: README.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+- To get `size` of the repository, the plugin uses [this](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#get-a-repository) endpoint.
```
Comment (narekhovhannisyan) [inline README.md]:
```suggestion
- To get `size` of the repository, the plugin uses [get a repository](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#get-a-repository) endpoint.
```

File: src/lib/github/api.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      );
+    }
+  };
+
+  /**
+   * Checks if the given `owner` is valid.
+   */
+  const checkOwnerValidity = async (owner: string) => {
+    try {
+      return await axios.get(`https://api.github.com/orgs/${owner}`);
```
Comment (narekhovhannisyan) [inline src/lib/github/api.ts]: please setup base axios object 

```ts
const GithubAPI = axios.createInstance({ baseURL: https://api.github.com/' })
```

and reuse it in methods

```ts
const checkOwnerValidity = async (owner: string) => {
    try {
      return await GithubAPI.get(`orgs/${owner}`);
      } ...
```

File: src/lib/github/api.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        `https://api.github.com/repos/${owner}/${repo}/traffic/clones`
+      );
+
+      if (result?.status !== 200) {
+        throw new APIRequestError(
+          `Error fetching clones from GitHub API for owner: ${owner} and repo: ${repo}. ${JSON.stringify(result?.status)}`
+        );
+      }
+
+      return result?.data.clones;
```
Comment (narekhovhannisyan) [inline src/lib/github/api.ts]: you can use response interceptor to make accessing the response data quicker

File: src/lib/github/index.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+import { validate } from '../../util/validations';
+
+import { GithubAPI } from './api';
+
+const { GlobalConfigError } = ERRORS;
+
+export const Github = (globalConfig: ConfigParams): ExecutePlugin => {
+  const metadata = {
+    kind: 'execute',
```
Comment (narekhovhannisyan) [inline src/lib/github/index.ts:19]: shouldn't we add input and output data to it?

File: src/util/errors.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
+const CUSTOM_ERRORS = ['AuthorizationError', 'APIRequestError'] as const;
+
+type CustomErrors = {
+  [K in (typeof CUSTOM_ERRORS)[number]]: ErrorConstructor;
+};
+
+export const ERRORS = CUSTOM_ERRORS.reduce((acc, className) => {
```
Comment (narekhovhannisyan) [inline src/util/errors.ts]: We have to use if-core

**Review by narekhovhannisyan — CHANGES_REQUESTED** [review]
Please check the comment above

**Review by jmcook1186 — APPROVED** [review]
sgtm!

**Review by narekhovhannisyan — APPROVED** [review]

**Comment by narekhovhannisyan** [comment]
@jmcook1186 It will be better to incapsulate validation into if-core too. We will have a lot of duplication across all plugin repos.

cc: @manushak

**Pull request merged**